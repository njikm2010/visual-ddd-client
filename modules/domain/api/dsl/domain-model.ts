import * as ViewDSL from '@/modules/domain/domain-design/dsl/dsl';
import { VoidClass, Void } from '@/modules/domain/domain-design/dsl/constants';
import { stringifyTypeDSL } from '@/modules/domain/domain-design/dsl/stringify';

import * as DSL from './interface';
import { ROOT, Tree, CoreProperties, BaseContainer, Node, IContainer } from './shared';

/**
 * 转换字符串，如果为空则返回 undefined
 * @param str
 * @returns
 */
export function transformString(str?: string): string | undefined {
  if (!str) {
    return;
  }

  const trimmed = str.trim();

  if (trimmed === '') {
    return;
  }

  return trimmed;
}

/**
 * 转换元数据
 * @param meta
 * @returns
 */
export function transformMeta(meta?: ViewDSL.MetaDSL[]): DSL.MetaDSL | undefined {
  if (!meta) {
    return;
  }

  return meta.reduce((acc, cur) => {
    if (cur.key && cur.value) {
      acc[cur.key] = cur.value;
    }

    return acc;
  }, {} as DSL.MetaDSL);
}

/**
 * 转换类型
 * @param type
 * @param getReference
 * @returns
 */
export function transformType(
  type: ViewDSL.TypeDSL,

  getReference: (id: string) => ViewDSL.NameDSL
): string {
  return stringifyTypeDSL(type, id => {
    const dsl = getReference(id);

    // 引用类型返回对应的对象的 [name:id]
    return `[${dsl.name}:${id}]`;
  });
}

export function transformProperty(
  property: ViewDSL.PropertyDSL,
  getReference: (id: string) => ViewDSL.NameDSL
): DSL.PropertyDSL {
  const { uuid, title, name, description, meta, access, type, optional } = property;

  return {
    uuid,
    title,
    name,
    description,
    meta: transformMeta(meta),
    optional,
    access,
    type: type ? transformType(type, getReference) : VoidClass,
  };
}

export function transformParameter(
  property: ViewDSL.ParameterDSL,
  getReference: (id: string) => ViewDSL.NameDSL
): DSL.ParameterDSL {
  const { uuid, title, name, description, meta, type, optional } = property;

  return {
    uuid,
    title,
    name,
    description,
    optional,
    meta: transformMeta(meta),
    type: type ? transformType(type, getReference) : VoidClass,
  };
}

export function transformMethods(
  methods: ViewDSL.MethodDSL[],
  getReference: (id: string) => ViewDSL.NameDSL
): DSL.MethodDSL[] {
  const map: Record<string, DSL.MethodDSL> = {};

  for (const m of methods) {
    const { uuid, title, name, description, meta, access, abstract, parameters, result } = m;

    const define = map[name];

    const signature: DSL.MethodDefineDSL = {
      description,
      parameters: parameters.map(p => transformParameter(p, getReference)),
      return: {
        type: result ? transformType(result, getReference) : Void,
      },
    };

    if (define) {
      define.signature = Array.isArray(define.signature) ? define.signature : [define.signature];
      define.signature.push(signature);
    } else {
      map[name] = {
        uuid,
        title,
        name,
        description,
        meta: transformMeta(meta),
        access,
        abstract,
        signature,
      };
    }
  }

  return Object.values(map);
}

export function transformEnum(e: ViewDSL.EnumDSL): DSL.EnumDSL {
  const { uuid, title, name, description, meta, baseType, members } = e;

  return {
    uuid,
    title,
    name,
    description,
    meta: transformMeta(meta),
    baseType,
    members: members.map(m => {
      const { uuid, title, name, description, meta, code } = m;

      return {
        uuid,
        title,
        name,
        description,
        meta: transformMeta(meta),
        code: baseType === 'number' ? parseFloat(code) : code,
      };
    }),
  };
}

export function transformRule(e: ViewDSL.RuleDSL): DSL.RuleDSL {
  const { uuid, title, name, description, meta } = e;

  return {
    uuid,
    title,
    name,
    description,
    meta: transformMeta(meta),
  };
}

export function transformSource(source: ViewDSL.SourceDSL): DSL.SourceDSL[] {
  const list: DSL.SourceDSL[] = [];
  if (source.http.enabled) {
    list.push({ type: 'http' });
  }
  if (source.rpc.enabled) {
    list.push({ type: 'rpc' });
  }
  if (source.event.enabled) {
    list.push({ type: 'event', value: source.event.value });
  }

  if (source.schedule.enabled) {
    list.push({ type: 'schedule', value: source.schedule.value });
  }

  return list;
}

/**
 * 实体
 */
export class Entity extends Node<ViewDSL.EntityDSL> {
  /**
   * 所属聚合
   */
  aggregation!: Aggregation;

  toDSL(): DSL.EntityDSL {
    const { uuid, title, name, description, meta, properties, methods, id } = this.properties;

    const propertiesDSL = properties.map(i => {
      return transformProperty(i, this.getReference);
    });

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      id: propertiesDSL.find(i => i.uuid === id)!,
      properties: propertiesDSL.filter(i => i.uuid !== id),
      methods: transformMethods(methods, this.getReference),
    };
  }
}

/**
 * 值对象
 */
export class ValueObject extends Node<ViewDSL.ValueObjectDSL> {
  /**
   * 所属聚合
   */
  aggregation!: Aggregation;

  toDSL(): DSL.ValueObjectDSL {
    const { uuid, title, name, description, meta, properties, methods } = this.properties;

    const propertiesDSL = properties.map(i => {
      return transformProperty(i, this.getReference);
    });

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      properties: propertiesDSL,
      methods: transformMethods(methods, this.getReference),
    };
  }
}

export class Enum extends Node<ViewDSL.EnumDSL> {
  /**
   * 所属聚合
   */
  aggregation!: Aggregation;

  toDSL(): DSL.EnumDSL {
    return transformEnum(this.properties);
  }
}

/**
 * 规则
 */
export class Rule extends Node<ViewDSL.RuleDSL> {
  toDSL(): DSL.RuleDSL {
    return transformRule(this.properties);
  }
}

export class Command extends Node<ViewDSL.CommandDSL> {
  rules: Rule[] = [];

  /**
   * 所属聚合
   */
  aggregation!: Aggregation;

  toDSL(): DSL.CommandDSL {
    const {
      uuid,
      title,
      name,
      description,
      meta,
      repository,
      eventSendable,
      category,
      result,
      properties,
      source,
      eventProperties,
    } = this.properties;

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      rules: this.rules.map(i => i.toDSL()),
      category: transformString(category),
      repository,
      eventSendable,
      source: transformSource(source),
      properties: properties.map(i => transformProperty(i, this.getReference)),
      event: {
        uuid,
        name: `${name}Event`,
        properties: eventProperties.map(i => transformProperty(i, this.getReference)),
      },
      return: {
        type: result ? transformType(result, this.getReference) : Void,
      },
    };
  }
}

export class Aggregation extends Node<ViewDSL.AggregationDSL> {
  private entities: Entity[] = [];
  private valueObjects: ValueObject[] = [];
  private enums: Enum[] = [];
  private commands: Command[] = [];

  addEntity(entity: Entity) {
    this.entities.push(entity);

    entity.aggregation = this;
  }

  addValueObject(valueObject: ValueObject) {
    this.valueObjects.push(valueObject);

    valueObject.aggregation = this;
  }

  addEnum(enumObject: Enum) {
    this.enums.push(enumObject);
    enumObject.aggregation = this;
  }

  addCommand(command: Command) {
    this.commands.push(command);
    command.aggregation = this;
  }

  /**
   * 转换为 dsl
   */
  toDSL(): DSL.AggregateDSL {
    const { uuid, title, name, description, meta } = this.properties;

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      root: this.entities.find(i => i.properties.isAggregationRoot)!.toDSL(),
      entities: this.entities.filter(i => !i.properties.isAggregationRoot).map(i => i.toDSL()),
      valueObjects: this.valueObjects.map(i => i.toDSL()),
      enums: this.enums.map(i => i.toDSL()),
      commands: this.commands.map(i => i.toDSL()),
    };
  }
}

export class DomainModelContainer extends BaseContainer implements IContainer {
  aggregations: Aggregation[] = [];

  /**
   * 收集所有有效的节点
   */
  nodesIndexByUUID: Map<string, Node> = new Map();

  /**
   * 深度递归遍历
   * @param tree
   */
  constructor(tree: Record<string, Tree>) {
    super();

    this.traverse(tree);
  }

  getNodeById(id: string): Node<any> | undefined {
    return this.nodesIndexByUUID.get(id);
  }

  /**
   * 转换为聚合 DSL
   * @returns
   */
  toDSL(): DSL.DomainModelDSL {
    return {
      aggregates: this.aggregations.map(i => i.toDSL()),
    };
  }

  handle(node: Tree, tree: Record<string, Tree>): void {
    const getParent = (n: Tree) => {
      const parentId = n.parent;
      if (parentId == null || parentId === ROOT) {
        return null;
      }

      return tree[parentId];
    };

    const isAggregation = (n: Tree): n is Tree<ViewDSL.AggregationDSL & CoreProperties> => {
      return n.properties.__node_name__ === 'aggregation';
    };

    const type = node.properties.__node_name__;

    if (node.properties.uuid && node.id !== node.properties.uuid) {
      throw new Error(`uuid 不匹配: ${node.id} ${JSON.stringify(node.properties)}`);
    }

    switch (type) {
      case 'aggregation': {
        const properties = node.properties as unknown as ViewDSL.AggregationDSL;
        const aggregation = new Aggregation(properties.uuid, properties, this);
        this.nodesIndexByUUID.set(aggregation.id, aggregation);

        this.aggregations.push(aggregation);

        // 聚合
        break;
      }
      case 'entity': {
        const properties = node.properties as unknown as ViewDSL.EntityDSL;
        const entity = new Entity(properties.uuid, properties, this);
        this.nodesIndexByUUID.set(entity.id, entity);

        const parent = getParent(node);

        if (parent && isAggregation(parent)) {
          const aggregation = this.nodesIndexByUUID.get(parent.properties.uuid) as Aggregation;
          aggregation.addEntity(entity);
        }

        break;
      }
      case 'value-object': {
        const properties = node.properties as unknown as ViewDSL.ValueObjectDSL;
        const valueObject = new ValueObject(properties.uuid, properties, this);
        this.nodesIndexByUUID.set(valueObject.id, valueObject);

        const parent = getParent(node);

        if (parent && isAggregation(parent)) {
          const aggregation = this.nodesIndexByUUID.get(parent.properties.uuid) as Aggregation;
          aggregation.addValueObject(valueObject);
        }

        break;
      }
      case 'enum': {
        const properties = node.properties as unknown as ViewDSL.EnumDSL;
        const enumObject = new Enum(properties.uuid, properties, this);
        this.nodesIndexByUUID.set(enumObject.id, enumObject);

        const parent = getParent(node);

        if (parent && isAggregation(parent)) {
          const aggregation = this.nodesIndexByUUID.get(parent.properties.uuid) as Aggregation;
          aggregation.addEnum(enumObject);
        }

        break;
      }
      case 'command': {
        const properties = node.properties as unknown as ViewDSL.CommandDSL;
        const command = new Command(properties.uuid, properties, this);

        this.nodesIndexByUUID.set(command.id, command);

        this.addPostTraverse(() => {
          if (properties.aggregation) {
            const aggregation = this.nodesIndexByUUID.get(properties.aggregation.referenceId) as
              | Aggregation
              | undefined;
            aggregation?.addCommand(command);
          }
        });
        break;
      }
      case 'rule': {
        const properties = node.properties as unknown as ViewDSL.RuleDSL;
        const rule = new Rule(properties.uuid, properties, this);

        this.nodesIndexByUUID.set(rule.id, rule);

        this.addPostTraverse(() => {
          if (properties.aggregator) {
            const aggregator = this.nodesIndexByUUID.get(properties.aggregator.referenceId) as Command | undefined;
            aggregator?.rules.push(rule);
          }
        });

        break;
      }
    }
  }
}
