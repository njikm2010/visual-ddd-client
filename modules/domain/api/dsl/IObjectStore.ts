import { Entity, Node, ValueObject } from './domain-model';
import { DTO } from './query-model';
import { DataObject } from './data-model';

export interface IObjectStore {
  getObjectById: (id: string) => { object: Entity | ValueObject | DTO | DataObject; parent?: Node };
}
