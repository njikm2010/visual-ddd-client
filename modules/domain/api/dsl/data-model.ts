/**
 * 数据模型 DSL 转换
 */
import * as DSL from './interface';
import * as ViewDSL from '@/modules/domain/data-design/dsl/dsl';
import { DataObjectName } from '@/modules/domain/data-design/dsl/constants';

import { transformMeta, transformString } from './domain-model';
import { Tree, BaseContainer, IContainer, Node } from './shared';

const DataObjectTypeName = ViewDSL.DataObjectTypeName;

export function transformType(
  type: ViewDSL.DataObjectTypeDSL,
  getReference: (objectId: string, propertyId: string) => ViewDSL.DataObjectPropertyDSL
): Record<string, any> {
  const typeName = type.type;

  switch (typeName) {
    case DataObjectTypeName.Date:
    case DataObjectTypeName.DateTime:
    case DataObjectTypeName.Timestamp:
      return {
        ...type,
        defaultValue: transformString(type.defaultValue),
      };
    case DataObjectTypeName.Reference:
      const refProperty = getReference(type.target!, type.targetProperty!);
      return {
        type: refProperty.type.type,
      };
    default:
      return type;
  }
}

export function transformProperty(
  property: ViewDSL.DataObjectPropertyDSL,
  getReference: (objectId: string, propertyId: string) => ViewDSL.DataObjectPropertyDSL
): DSL.DataPropertyDSL {
  const { uuid, title, name, description, meta, propertyName, primaryKey, notNull, type } = property;

  return {
    uuid,
    name,
    title,
    description,
    meta: transformMeta(meta),
    propertyName: transformString(propertyName),
    ...(transformType(type, getReference) as {
      type: DSL.DataPropertyType;
      defaultValue?: any;
      autoIncrement?: boolean;
      precision?: number;
      scale?: number;
      length?: number;
    }),
    notNull: primaryKey ? true : notNull,
  };
}

export function transformIndex(
  index: ViewDSL.DataObjectIndexDSL,
  getProperty: (id: string) => ViewDSL.DataObjectPropertyDSL
): DSL.DataIndexDSL {
  const { uuid, title, name, description, meta, type, method, properties } = index;

  return {
    uuid,
    name,
    title,
    description,
    meta: transformMeta(meta),
    type,
    method,
    properties: properties.map(i => getProperty(i).name),
  };
}

export function transformDataObjectReference(
  viewDSL: ViewDSL.DataObjectDSL,
  getObject: (objectId: string) => ViewDSL.DataObjectDSL,
  getReference: (objectId: string, propertyId: string) => ViewDSL.DataObjectPropertyDSL
): DSL.DataObjectReferenceDSL {
  const properties = viewDSL.properties;
  const targets: Map<
    // 目标对象 name
    string,
    {
      cardinality: 'OneToOne' | 'OneToMany' | 'ManyToMany' | 'ManyToOne';
      // 字段映射
      mapper: Map<string, string>;
    }
  > = new Map();

  for (const property of properties) {
    if (property.type.type !== DataObjectTypeName.Reference) {
      continue;
    }

    const { target, targetProperty, cardinality } = property.type as Required<ViewDSL.DataObjectReference>;

    // 自我引用去掉?
    if (target === viewDSL.uuid) {
      continue;
    }

    const targetObject = getObject(target);
    const targetPropertyInfo = getReference(target, targetProperty);
    const targetName = targetObject.name;

    let state = targets.get(targetName);

    if (state == null) {
      state = {
        cardinality,
        mapper: new Map(),
      };
      targets.set(targetName, state);
    }

    state.cardinality ??= cardinality;
    state.mapper.set(property.name, targetPropertyInfo.name);
  }

  return {
    source: viewDSL.name,
    targets: Array.from(targets.entries()).map(([target, { cardinality, mapper }]) => ({
      target,
      cardinality,
      mapper: Array.from(mapper.entries()).map(([sourceField, targetField]) => ({
        sourceField,
        targetField,
      })),
    })),
  };
}

export class DataObject extends Node<ViewDSL.DataObjectDSL> {
  toDSL(): DSL.DataObjectDSL {
    const { uuid, title, name, description, meta, properties, tableName, indexes } = this.properties;

    const transformedIndexes = indexes.map(i => transformIndex(i, this.getProperty));

    // 收集主键索引
    const primaryKeys = this.properties.properties.filter(i => i.primaryKey);
    if (primaryKeys.length) {
      transformedIndexes.push({
        uuid: `${uuid}-primary-key`,
        name: 'primaryKeys',
        type: 'Primary',
        properties: primaryKeys.map(i => i.name),
      });
    }

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      tableName: transformString(tableName),
      properties: properties.map(i => transformProperty(i, this.getPropertyReference)),
      indexes: transformedIndexes,
    };
  }

  toReferenceDSL(): DSL.DataObjectReferenceDSL {
    return transformDataObjectReference(this.properties, this.getObject, this.getPropertyReference);
  }

  getObject = (id: string) => {
    return this.getReference(id) as ViewDSL.DataObjectDSL;
  };

  getProperty = (id: string): ViewDSL.DataObjectPropertyDSL => {
    return this.properties.properties.find(i => i.uuid === id)!;
  };

  getPropertyReference = (id: string, propertyId: string) => {
    const node = this.container.getNodeById(id) as unknown as DataObject;

    return node.properties.properties.find(i => i.uuid === propertyId)!;
  };
}

/**
 * 数据对象容器
 */
export class DataModelContainer extends BaseContainer implements IContainer {
  dataObjects: DataObject[] = [];

  /**
   * 收集所有有效的节点
   */
  nodesIndexByUUID: Map<string, DataObject> = new Map();

  constructor(tree: Record<string, Tree>) {
    super();

    this.traverse(tree);
  }

  toDSL(): DSL.DataModelDSL {
    return {
      dataObjects: this.dataObjects.map(i => i.toDSL()),
      references: this.dataObjects.map(i => i.toReferenceDSL()).filter(i => i.targets.length),
    };
  }

  getNodeById(id: string): Node<any> | undefined {
    return this.nodesIndexByUUID.get(id);
  }

  handle(node: Tree, tree: Record<string, Tree>): void {
    const type = node.properties.__node_name__;

    if (type !== DataObjectName.DataObject) {
      return;
    }

    const properties = node.properties as unknown as ViewDSL.DataObjectDSL;
    const object = new DataObject(properties.uuid, properties, this);
    this.dataObjects.push(object);
    this.nodesIndexByUUID.set(object.id, object);
  }
}
