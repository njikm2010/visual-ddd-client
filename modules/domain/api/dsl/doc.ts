import { Doc as YDoc, applyUpdate } from 'yjs';

import { YJS_FIELD_NAME } from '../../constants';

import * as DSL from './interface';
import { transform as transformToDomainDSL } from './domain-model';
import { transform as transformToQueryDSL } from './query-model';
import { transform as transformToDataObjectDSL } from './data-model';

/**
 * 创建文档
 * @param update
 * @returns
 */
export function createDoc(update: Uint8Array) {
  const doc = new YDoc();

  applyUpdate(doc, update);

  return doc;
}

/**
 * 转换为 DSL
 * @param doc
 * @returns
 */
export function transformToDSL(doc: YDoc): DSL.BusinessDomainDSL {
  const domainMap = doc.getMap(YJS_FIELD_NAME.DOMAIN);
  const queryMap = doc.getMap(YJS_FIELD_NAME.QUERY);
  const dataObjectMap = doc.getMap(YJS_FIELD_NAME.DATA_OBJECT);

  return {
    domainModel: transformToDomainDSL(domainMap.toJSON()),
    queryModel: transformToQueryDSL(queryMap.toJSON()),
    dataModel: transformToDataObjectDSL(dataObjectMap.toJSON()),
  };
}
