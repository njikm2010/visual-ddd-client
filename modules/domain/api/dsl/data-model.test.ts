import {
  transformProperty,
  transformType,
  transformIndex,
  transformDataObjectReference,
  transform,
} from './data-model';
import * as ViewDSL from '@/modules/domain/data-design/dsl/dsl';

const getReference = (id: string, propertyId: string): ViewDSL.DataObjectPropertyDSL => {
  return {
    uuid: propertyId,
    name: 'name',
    type: { type: ViewDSL.DataObjectTypeName.Date },
  };
};

test('transformType', () => {
  expect(
    transformType(
      {
        type: ViewDSL.DataObjectTypeName.Date,
        // @ts-expect-error
        defaultValue: '',
      },
      getReference
    )
  ).toEqual({
    type: ViewDSL.DataObjectTypeName.Date,
    defaultValue: undefined,
  });

  expect(
    transformType(
      {
        type: ViewDSL.DataObjectTypeName.Boolean,
        defaultValue: true,
      },
      getReference
    )
  ).toEqual({
    type: ViewDSL.DataObjectTypeName.Boolean,
    defaultValue: true,
  });

  expect(
    transformType(
      {
        type: ViewDSL.DataObjectTypeName.Reference,
        target: 'target',
        targetProperty: 'targetProperty',
      },
      getReference
    )
  ).toEqual({
    type: ViewDSL.DataObjectTypeName.Date,
  });
});

test('transformProperty', () => {
  expect(
    transformProperty(
      {
        uuid: 'uuid',
        name: 'name',
        title: 'title',
        description: 'description',
        meta: [],
        propertyName: '',
        type: {
          type: ViewDSL.DataObjectTypeName.Date,
          defaultValue: '2022-12-12',
        },
        primaryKey: true,
      },
      getReference
    )
  ).toEqual({
    defaultValue: '2022-12-12',
    description: 'description',
    meta: {},
    name: 'name',
    notNull: true,
    propertyName: undefined,
    title: 'title',
    type: 'Date',
    uuid: 'uuid',
  });
});

test('transformIndex', () => {
  expect(
    transformIndex(
      {
        uuid: 'uuid',
        name: 'name',
        title: 'title',
        description: 'description',
        meta: [],
        properties: ['1', '2'],
        type: ViewDSL.DataObjectIndexType.Unique,
        method: ViewDSL.DataObjectIndexMethod.HASH,
      },
      id => {
        return { uuid: id, name: `Name${id}`, type: { type: ViewDSL.DataObjectTypeName.Date } };
      }
    )
  ).toEqual({
    description: 'description',
    meta: {},
    method: 'HASH',
    name: 'name',
    properties: ['Name1', 'Name2'],
    title: 'title',
    type: 'Unique',
    uuid: 'uuid',
  });
});

test('transformDataObjectReference', () => {
  expect(
    transformDataObjectReference(
      {
        uuid: 'me',
        name: 'name',
        title: 'title',
        properties: [
          {
            uuid: 'uuid',
            name: 'one',
            type: {
              type: ViewDSL.DataObjectTypeName.Reference,
              target: 'target',
              targetProperty: 'one',
              cardinality: ViewDSL.DataObjectReferenceCardinality.ManyToOne,
            },
          },
          {
            uuid: 'uuid',
            name: 'two',
            type: {
              type: ViewDSL.DataObjectTypeName.Reference,
              target: 'target',
              targetProperty: 'two',
            },
          },
          // 自我引用
          {
            uuid: 'uuid',
            name: 'three',
            type: {
              type: ViewDSL.DataObjectTypeName.Reference,
              target: 'me',
              targetProperty: 'two',
            },
          },
        ],
        indexes: [],
      },
      id => {
        return {
          uuid: id,
          name: `name${id}`,
          properties: [],
          indexes: [],
        };
      },
      (id, propertyId) => {
        return {
          uuid: propertyId,
          name: `name${propertyId}`,
          type: { type: ViewDSL.DataObjectTypeName.Date },
        };
      }
    )
  ).toEqual({
    source: 'name',
    targets: [
      {
        target: 'nametarget',
        cardinality: 'ManyToOne',
        mapper: [
          {
            sourceField: 'one',
            targetField: 'nameone',
          },
          {
            sourceField: 'two',
            targetField: 'nametwo',
          },
        ],
      },
    ],
  });
});

test('transform', () => {
  const data = {
    __ROOT__: {
      __NODE__: true,
      id: '__ROOT__',
      children: {
        'b51bebe2-85ee-4733-b9a0-25926d3d93dc': 1,
        '28d7cc24-2bef-49fe-a4a1-7e3e11c87796': 1,
        'copy-f2853107-0547-4ade-bf42-90ec5cbfcb24': 1,
      },
      properties: { __node_name__: '__ROOT__', __node_type__: 'node', __PROPERTY__: true },
    },
    'b51bebe2-85ee-4733-b9a0-25926d3d93dc': {
      __NODE__: true,
      id: 'b51bebe2-85ee-4733-b9a0-25926d3d93dc',
      parent: '__ROOT__',
      children: {},
      properties: {
        __node_name__: 'dataObject',
        __node_type__: 'node',
        position: { x: 60, y: 45 },
        uuid: 'b51bebe2-85ee-4733-b9a0-25926d3d93dc',
        name: 'Untitled',
        title: '未命名',
        description: '',
        meta: [],
        properties: [
          {
            uuid: '237497a4-0497-4088-b4a6-1fadc7a24987',
            name: 'a',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: false,
            type: {
              type: 'Reference',
              target: '28d7cc24-2bef-49fe-a4a1-7e3e11c87796',
              targetProperty: '4338a498-08c1-45a3-b0de-8ba7b2d28a42',
              cardinality: 'OneToMany',
            },
          },
          {
            uuid: '49d8636d-71f9-4e1e-aa34-c68e5bec4be1',
            name: 'b',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: false,
            type: {
              type: 'Reference',
              target: '28d7cc24-2bef-49fe-a4a1-7e3e11c87796',
              targetProperty: '8ade40eb-c860-42e6-9093-413ffb8972e4',
              cardinality: 'ManyToMany',
            },
          },
          {
            uuid: '4cbb88ae-8c97-482f-bfbf-9376201447e7',
            name: 'name',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: true,
            type: { type: 'String' },
          },
        ],
        indexes: [
          {
            uuid: 'd4d04958-b0a7-4967-bdc0-c4e7444eccb0',
            name: 'untitled0',
            title: '未命名',
            description: '',
            meta: [],
            type: 'Unique',
            properties: ['237497a4-0497-4088-b4a6-1fadc7a24987', '49d8636d-71f9-4e1e-aa34-c68e5bec4be1'],
            method: 'BTREE',
          },
        ],
        zIndex: 2,
        __PROPERTY__: true,
        size: { width: 200, height: 118.1875 },
      },
    },
    '28d7cc24-2bef-49fe-a4a1-7e3e11c87796': {
      __NODE__: true,
      id: '28d7cc24-2bef-49fe-a4a1-7e3e11c87796',
      parent: '__ROOT__',
      children: {},
      properties: {
        __node_name__: 'dataObject',
        __node_type__: 'node',
        position: { x: 60, y: 240 },
        uuid: '28d7cc24-2bef-49fe-a4a1-7e3e11c87796',
        name: 'HelloWorld',
        title: '你好世界',
        description: '',
        meta: [],
        properties: [
          {
            uuid: '4338a498-08c1-45a3-b0de-8ba7b2d28a42',
            name: 'untitled',
            title: '',
            description: '',
            meta: [],
            notNull: true,
            primaryKey: false,
            type: { type: 'String' },
          },
          {
            uuid: '8ade40eb-c860-42e6-9093-413ffb8972e4',
            name: 'untitled1',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: true,
            type: { type: 'String' },
          },
          {
            uuid: 'e9bbcdbd-15ec-49db-9130-d0069f89f3ef',
            name: 'untitled2',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: false,
            type: { type: 'String' },
          },
        ],
        indexes: [],
        zIndex: 2,
        __PROPERTY__: true,
        size: { width: 200, height: 118.1875 },
      },
    },
    'copy-f2853107-0547-4ade-bf42-90ec5cbfcb24': {
      __NODE__: true,
      id: 'copy-f2853107-0547-4ade-bf42-90ec5cbfcb24',
      parent: '__ROOT__',
      children: {},
      properties: {
        __node_name__: 'dataObject',
        __node_type__: 'node',
        uuid: 'copy-f2853107-0547-4ade-bf42-90ec5cbfcb24',
        name: 'Just',
        title: '你好世界',
        description: '',
        meta: [],
        properties: [
          {
            uuid: '4338a498-08c1-45a3-b0de-8ba7b2d28a42',
            name: 'untitled',
            title: '',
            description: '',
            meta: [],
            notNull: true,
            primaryKey: false,
            type: { type: 'String', defaultValue: 'hello', length: 6 },
          },
          {
            uuid: '8ade40eb-c860-42e6-9093-413ffb8972e4',
            name: 'untitled1',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: true,
            type: { type: 'String' },
          },
          {
            uuid: 'e9bbcdbd-15ec-49db-9130-d0069f89f3ef',
            name: 'untitled2',
            title: '',
            description: '',
            meta: [],
            notNull: false,
            primaryKey: false,
            type: { type: 'String' },
          },
        ],
        indexes: [],
        zIndex: 2,
        position: { x: 302, y: 143.1875 },
        __PROPERTY__: true,
        size: { width: 200, height: 118.1875 },
      },
    },
  };

  expect(transform(data as any)).toMatchSnapshot();
});
