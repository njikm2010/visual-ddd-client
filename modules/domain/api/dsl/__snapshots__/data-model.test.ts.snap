// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform 1`] = `
{
  "dataObjects": [
    {
      "description": "",
      "indexes": [
        {
          "description": "",
          "meta": {},
          "method": "BTREE",
          "name": "untitled0",
          "properties": [
            "a",
            "b",
          ],
          "title": "未命名",
          "type": "Unique",
          "uuid": "d4d04958-b0a7-4967-bdc0-c4e7444eccb0",
        },
        {
          "name": "primaryKeys",
          "properties": [
            "name",
          ],
          "type": "Primary",
          "uuid": "b51bebe2-85ee-4733-b9a0-25926d3d93dc-primary-key",
        },
      ],
      "meta": {},
      "name": "Untitled",
      "properties": [
        {
          "description": "",
          "meta": {},
          "name": "a",
          "notNull": false,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "237497a4-0497-4088-b4a6-1fadc7a24987",
        },
        {
          "description": "",
          "meta": {},
          "name": "b",
          "notNull": false,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "49d8636d-71f9-4e1e-aa34-c68e5bec4be1",
        },
        {
          "description": "",
          "meta": {},
          "name": "name",
          "notNull": true,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "4cbb88ae-8c97-482f-bfbf-9376201447e7",
        },
      ],
      "tableName": undefined,
      "title": "未命名",
      "uuid": "b51bebe2-85ee-4733-b9a0-25926d3d93dc",
    },
    {
      "description": "",
      "indexes": [
        {
          "name": "primaryKeys",
          "properties": [
            "untitled1",
          ],
          "type": "Primary",
          "uuid": "28d7cc24-2bef-49fe-a4a1-7e3e11c87796-primary-key",
        },
      ],
      "meta": {},
      "name": "HelloWorld",
      "properties": [
        {
          "description": "",
          "meta": {},
          "name": "untitled",
          "notNull": true,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "4338a498-08c1-45a3-b0de-8ba7b2d28a42",
        },
        {
          "description": "",
          "meta": {},
          "name": "untitled1",
          "notNull": true,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "8ade40eb-c860-42e6-9093-413ffb8972e4",
        },
        {
          "description": "",
          "meta": {},
          "name": "untitled2",
          "notNull": false,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "e9bbcdbd-15ec-49db-9130-d0069f89f3ef",
        },
      ],
      "tableName": undefined,
      "title": "你好世界",
      "uuid": "28d7cc24-2bef-49fe-a4a1-7e3e11c87796",
    },
    {
      "description": "",
      "indexes": [
        {
          "name": "primaryKeys",
          "properties": [
            "untitled1",
          ],
          "type": "Primary",
          "uuid": "copy-f2853107-0547-4ade-bf42-90ec5cbfcb24-primary-key",
        },
      ],
      "meta": {},
      "name": "Just",
      "properties": [
        {
          "defaultValue": "hello",
          "description": "",
          "length": 6,
          "meta": {},
          "name": "untitled",
          "notNull": true,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "4338a498-08c1-45a3-b0de-8ba7b2d28a42",
        },
        {
          "description": "",
          "meta": {},
          "name": "untitled1",
          "notNull": true,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "8ade40eb-c860-42e6-9093-413ffb8972e4",
        },
        {
          "description": "",
          "meta": {},
          "name": "untitled2",
          "notNull": false,
          "propertyName": undefined,
          "title": "",
          "type": "String",
          "uuid": "e9bbcdbd-15ec-49db-9130-d0069f89f3ef",
        },
      ],
      "tableName": undefined,
      "title": "你好世界",
      "uuid": "copy-f2853107-0547-4ade-bf42-90ec5cbfcb24",
    },
  ],
  "references": [
    {
      "source": "Untitled",
      "sourceId": "b51bebe2-85ee-4733-b9a0-25926d3d93dc",
      "targets": [
        {
          "cardinality": "OneToMany",
          "mapper": [
            {
              "sourceField": "a",
              "targetField": "untitled",
            },
            {
              "sourceField": "b",
              "targetField": "untitled1",
            },
          ],
          "target": "HelloWorld",
          "targetId": "28d7cc24-2bef-49fe-a4a1-7e3e11c87796",
        },
      ],
    },
  ],
}
`;
